server/src/main/java/io/crate/planner/optimizer/rule/FilterOnJoinsUtil.java

```diff
@@ -60,11 +60,37 @@ static LogicalPlan moveQueryBelowJoin(Symbol query, JoinPlan join) {
         }
         LogicalPlan lhs = join.lhs();
         LogicalPlan rhs = join.rhs();
-        Symbol queryForLhs = splitQuery.remove(new HashSet<>(lhs.getRelationNames()));
-        Symbol queryForRhs = splitQuery.remove(new HashSet<>(rhs.getRelationNames()));
-        LogicalPlan newLhs = getNewSource(queryForLhs, lhs);
-        LogicalPlan newRhs = getNewSource(queryForRhs, rhs);
-        LogicalPlan newJoin = join.replaceSources(List.of(newLhs, newRhs));
+
+        HashSet<RelationName> lhsRelations = new HashSet<>(lhs.getRelationNames());
+        HashSet<RelationName> rhsRelations = new HashSet<>(rhs.getRelationNames());
+
+        Symbol queryForLhs = splitQuery.remove(lhsRelations);
+        Symbol queryForRhs = splitQuery.remove(rhsRelations);
+
+        LogicalPlan newLhs = null;
+        LogicalPlan newRhs = null;
+
+        if (queryForRhs != null || queryForLhs != null) {
+            newLhs = getNewSource(queryForLhs, lhs);
+            newRhs = getNewSource(queryForRhs, rhs);
+        } else {
+            for (var relationNames : Set.copyOf(splitQuery.keySet())) {
```

Comment:
The set is copied to be able to iterate and modify splitQuery at the same time.

Reply:
Question: is order of iteration over `splitQuery.keySet()` important?

Just have a concern, that `splitQuery.keySet()` seems to be backed by a `LinkedHashMap`, i.e has concrete order iteration semantics and `Set.copyOf` looks like `Set.of(new HashSet<>(coll).toArray());` for mutable  collections.

Reply:
I guess if the filter matches multiple sides of the join the behaviour could become flaky something like:

```
Filter[(d > 1)]
 └ HashJoin[(c = d)]                
    ├ HashJoin[(a = b)]        
    │  ├ Collect[doc.t4 | [a] | true] 
    │  └ HashJoin[(d = c)] 
    |    ├ Collect[doc.t1 | [d] | true] 
    |    └ Collect[doc.t3 | [c] | true]    
    └ HashJoin[(d = e)] 
       ├ Collect[doc.t1 | [d] | true] 
       └ Collect[doc.t5 | [e] | true]   
```
Then the filter matches both side of the top join and the order matters. No sure, how realistic this is.

Reply:
Wouldn't it be more efficient to use an `splitQuery.entrySet().iterator` instead of copying the keySet?

Reply:
If you iterate over a map and modify it with the same iterator you get a `ConcurrentModificationException`. This is why it's copied.